.github/workflows/tests.yaml

```diff
@@ -0,0 +1,31 @@
+name: Unit and Integration Test
+
+on:
+  push:
+    branches: [ test-build-github-actions ]
+  pull_request:
+    branches: [ master ]
+
+env:
+  CARGO_TERM_COLOR: always
+
+jobs:
+  test:
+    runs-on: ubuntu-latest
+    
+    steps:
+    - uses: actions/checkout@v2
+    - name: Get Dependencies
+      run: |
+        sudo apt-get update
+        sudo apt-get install capnproto
+    - uses: actions-rs/toolchain@v1
+      with:
+        toolchain: stable
+    - uses: Swatinem/rust-cache@v1
+    - name: Unit Testing
+      run: cargo test --bins
+    - name: Config FUSE to run without sudo
+      run: sudo chown -R  $USER:$USER /etc/fuse.conf && sudo echo user_allow_other >> /etc/fuse.conf
```

Comment:
I wanted to comment on this line, but I forgot. I understand you are trying to write the 'user_allow_other' in the /etc/fuse.conf.

But, you should not change the ownership of the /etc/fuse.conf file. Most files under /etc/ should be owned by the root user.

I know you changed this so the `echo >> /etc/fuse.conf` work. right? but it's better to understand why it didn't work although you are already doing `sudo echo` in the first place. (hence you had to change the ownership to user $USER)

Okay, the first question is what does `echo` do. Echo is a very simple command that takes all it's arguments, and writes it to it's stdout. So what is stdout? stdout are one of the main open files that any process have open on start. Each process in linux when it starts it already have `stdin`, `stdout`, and `stderr`. The stdout is the one used by default when u do something like `println!()` in rust, stderr is the one that is used when u do `eprintln!()`. 

So what echo does, it takes each input from cmdline arguments, and write it out to stdout. Next question is where does stdout comes from. There can be many ways to pass the stdout to the process but in our case it's the shell that does it with the redirect operation `>>`

When you start a command on the shell and then do a redirect what happens the shell first open that file (the one that comes after `>` or `>>` and then set it as the stdout of the command that you are running.

so a simple command like
```
echo hello world > somefile.text
```
will make the shell first open the somefile.text in write mode, create the echo process and then assign it's stdout to that file, then start the process. so the echo will just still do `println!()` but because it's stdout now is actually the `somefile.text` the output from the print statement will go to that file.

Okay, great. Now what happens when u do this 
```
sudo echo hello world > somefile.text
``` 
Well, basically the same the shell will open the file, set the stdout of the executable you are trying to run. The only difference here is the command is actually `sudo` not `echo`. Sudo will then change it's active user to root (if you have permissions) and then start the command passed to it `echo` and pass the stdout to it so now echo is running as root and have the somefile.text as stdout. Yet it doesn't work, echo fails to write to that files. Why?

Because the file was actually open by the shell, not by the echo or the sudo command. The shell runs as a normal user, not root. hence what actually happen is the following

Shell opens the files (as user, in write mode) already the shell can't write because the file is owned by `root`. Then passes it as stdout to the sub-process (sudo then echo) echo tries to write that was opened by user hence when it writes, the file can't be written to because whoever opened it does not have permissions to do so.

What is the solution to this. Simple, open the file as root then echo won't have issue writing to it (even without sudo)

How?

```suggestion
      run:  echo user_allow_other | sudo tee /etc/fuse.conf
```
So what is `tee` command. `tee` is another tool (like echo) that takes input from it's `stdin` and write it to the file given via the argument.
So `tee` is going to open the file itself, hence will open it with its own permissions. so  a `sudo tee` will have the file open as `root`. Then it will take what is written to it via it's stdin (in that case the pipe from echo) and write it to that file hence it will succeed to do that

So on running the above the command that's what will happen.
Shell will create two processes:
- on the left side, create an echo process.
- on the right side it will create a `sudo` process.
- it will create a pipe (that's special files in linux you can read about it in the book i gave you)
- it will then set one end of the pipe (the write end of the pipe) as the echo stdout. So when echo do a `println!()` it will end up in the pipe.
- The other end (the read end of the pipe) is attached to the sudo process stdin (and will eventually be passed to `tee` 
- then shell is going to actually run the two processes after setup is complete
- echo will just do it's print thingy, it will be written to the pipe to appear at tee stding, tee will read that and write it to the file.
- done.
-